### PR TITLE
D9-07: Send mail to registered user.

### LINKED ISSUE #11: #07 ­- Cron Queuing
## Objective
At the end of this course, you will be able to

- Understand and utilize the Queue API to add tasks to the queue and process them on cron runs
## Exercise
- Build a module that sends a welcome email to registered users
- Use the Cron Queue by adding to the queue the user id whenever a user registers
- Create a queue worker that picks these queue items (uids) during the cron and sends a welcome email (can be any simple text) to the registered email address.

### PR DESCRIPTION
**Pull request by amansrivastava**
#11

### CODE REVIEW

File: docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+   */
+  public static function create(
+    ContainerInterface $container,
+    array $configuration,
+    $plugin_id,
+    $plugin_definition
+  ) {
+    $mailManager = \Drupal::service('plugin.manager.mail');
+    $entityTypeManager = \Drupal::service('entity_type.manager');
+    return new static($mailManager, $entityTypeManager);
```
Comment (joshirohit100) [inline docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php]: this is not correct way to instantiate object via factory method. Pls check examples in core.

File: docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function processItem($userId) {
+    /** @var \Drupal\user\Entity\User $user */
+    $user = $this->entityTypeManager->getStorage('user')->load($userId);
+    $config = \Drupal::config("system.site");
```
Comment (joshirohit100) [inline docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php]: Inject the dependency.

File: docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      'user_mail',
+      'welcome_mail',
+      $user->getEmail(),
+      $user->getPreferredLangcode(),
+      ['message' => 'Welcome to ' . $config->get("name") . ', user ' . $user->getAccountName()],
+      NULL,
+      TRUE
+    );
+
+    $logger = \Drupal::logger('user_mail');
```
Comment (joshirohit100) [inline docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php]: Same here - Inject dependency.

File: docroot/modules/custom/user_mail/user_mail.services.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+services:
+  logger.channel.user_mail:
+    class: Drupal\Core\Logger\LoggerChannel
+    factory: logger.factory:get
+    arguments: ['user_mail']
```
Comment (joshirohit100) [inline docroot/modules/custom/user_mail/user_mail.services.yml:5]: where are we using this?

Comment (amansrivastava) [reply docroot/modules/custom/user_mail/user_mail.services.yml:5]: in WelcomeMail.php:83. 
https://github.com/amansrivastava/d9-dev/pull/28/files#diff-a5e1492db63f9eec8c3c9d419b31b86a8c683f26a891bca1328f185e27a34276R83

File: docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * @QueueWorker(
+ *   id = "user_mail_welcomemail",
+ *   title = @Translation("WelcomeMail"),
+ *   cron = {"time" = 60}
+ * )
+ */
+class WelcomeMail extends QueueWorkerBase implements ContainerFactoryPluginInterface {
+
+
```
Comment (joshirohit100) [inline docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php]: un-wanted extra line

File: docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    $this->configFactory = $config_factory;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('logger.factory')->get('user_mail'),
```
Comment (joshirohit100) [inline docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php:82]: why not using the logger channel created?

File: docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      NULL,
+      TRUE
+    );
+
+    if ($mail['result'] !== TRUE) {
+      $this->logger->error('There was a problem sending the mail to ' . $user->getEmail());
+    }
+    else {
+      $this->logger->info('The mail has been sent to ' . $user->getEmail());
+    }
```
Comment (joshirohit100) [inline docroot/modules/custom/user_mail/src/Plugin/QueueWorker/WelcomeMail.php]: we don;t need `else`, we can just return from `IF` if fail. If not fail, we just log success message without else.

**Pull request merged**